File: Middle of The Linked List/Go/solution2.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func middleNodeFastSlow(head *ListNode) *ListNode {
	// track a slow and fast node
	// slow advances one at a time
	// fast advances 2 nodes at a time
	slow, fast := head, head.Next

	// traverse the list until fast is either at the last node or past the last node
	for fast != nil && fast.Next != nil {
		// Advance both nodes as noted above
		slow = slow.Next
		fast = fast.Next.Next
	}
	// return the slow node
	return slow
}

File: Reverse Linked List/JavaScript/solution2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var reverseList = function(head) {
    // track the last node we saw
    prev = null
    // loop until we go past the new head
    while(head != null) {
        // save the current node
        let cur = head
        // the next node we visit will be the current node's next
        head = head.next
        // set the current node to point to the last node we saw
        cur.next = prev
        // save the node we are at as the previous node for next iterations
        prev = cur
    }
    
    return prev
};